app/models/user_password.rb

```diff
@@ -48,39 +79,25 @@ def expired?
     created_at < (Time.now - days_valid)
   end
 
-  # Returns a 128bits random salt as a hex string (32 chars long)
-  def self.generate_salt
-    SecureRandom.hex(16)
-  end
+  protected
 
-  # Return password digest
-  def self.hash_password(plain_password)
-    Digest::SHA1.hexdigest(plain_password)
+  # Save hashed_password from the initially passed plain password
+  # if it is is set.
+  def salt_and_hash_password!
+    return if plain_password.nil?
+    self.hashed_password = derive_password!(plain_password)
   end
 
-  # Hash a plaintext password with a given salt
-  # The hashed password has following form: SHA1(salt + SHA1(password))
-  def self.hash_with_salt(plain_password, salt)
-    # We should really use a standard key-derivation function like bcrypt here
-    hash_password("#{salt}#{hash_password plain_password}")
+  # Require the implementation to provide a secure comparisation
+  def hash_matches?(_plain)
+    raise NotImplementedError, 'Must be overridden by subclass'
   end
 
-  # constant-time comparison algorithm to prevent timing attacks
-  def self.secure_equals?(a, b)
-    return false if a.blank? || b.blank? || a.bytesize != b.bytesize
-    l = a.unpack "C#{a.bytesize}"
-
-    res = 0
-    b.each_byte do |byte| res |= byte ^ l.shift end
-    res == 0
+  def generate_salt
```

Comment:
Why is this still here? As far as I can tell this is no longer used anywhere. Also it's not implemented by either subclass.